File: src/pipeline/mesh-building/WindmeshFile.ts
/**
 * Binary .windmesh file format for prebuilt wind mesh data.
 *
 * Version 2 format (multi-source):
 *   Header (32 bytes):
 *     [0..3]   magic: "WNDM" (ASCII LE)
 *     [4..5]   version: u16 = 2
 *     [6..7]   sourceCount: u16
 *     [8..11]  inputHashHi: u32
 *     [12..15] inputHashLo: u32
 *     [16..31] reserved
 *
 *   Per-source direction table (4 bytes x sourceCount):
 *     direction: f32 (radians)
 *
 *   Per-source entry table (16 bytes x sourceCount):
 *     vertexDataOffset, vertexCount, indexDataOffset, indexCount (all u32)
 *
 *   Shared grid metadata (32 bytes):
 *     gridCols, gridRows, gridMinX, gridMinY, gridCellWidth, gridCellHeight, reserved x2
 *
 *   Data sections (concatenated):
 *     Vertex data per source (5 f32/vertex)
 *     Index data (shared)
 */

const MAGIC = 0x4d444e57; // "WNDM" as little-endian u32
const V2_HEADER_BYTES = 32;

/**
 * Per-source wind mesh data.
 */
export interface WindMeshSourceData {
  direction: number;
  vertices: Float32Array;
  indices: Uint32Array;
  vertexCount: number;
  indexCount: number;
}

/**
 * Multi-source wind mesh bundle.
 */
export interface WindMeshFileBundle {
  sources: WindMeshSourceData[];
  sourceCount: number;
  gridCols: number;
  gridRows: number;
  gridMinX: number;
  gridMinY: number;
  gridCellWidth: number;
  gridCellHeight: number;
}

/**
 * Legacy single-source data (kept for backward compat with version 1).
 */
export interface WindMeshFileData {
  vertices: Float32Array;
  indices: Uint32Array;
  vertexCount: number;
  indexCount: number;
  gridCols: number;
  gridRows: number;
  gridMinX: number;
  gridMinY: number;
  gridCellWidth: number;
  gridCellHeight: number;
}

function parseV1(buffer: ArrayBuffer): WindMeshFileBundle {
  const view = new DataView(buffer);
  const headerBytes = 16;
  const metadataBytes = 32;

  const metaOffset = headerBytes;
  const vertexCount = view.getUint32(metaOffset, true);
  const indexCount = view.getUint32(metaOffset + 4, true);
  const gridCols = view.getUint32(metaOffset + 8, true);
  const gridRows = view.getUint32(metaOffset + 12, true);
  const gridMinX = view.getFloat32(metaOffset + 16, true);
  const gridMinY = view.getFloat32(metaOffset + 20, true);
  const gridCellWidth = view.getFloat32(metaOffset + 24, true);
  const gridCellHeight = view.getFloat32(metaOffset + 28, true);

  const dataStart = headerBytes + metadataBytes;
  const vertices = new Float32Array(buffer, dataStart, vertexCount * 5);
  const indexStart = dataStart + vertexCount * 5 * 4;
  const indices = new Uint32Array(buffer, indexStart, indexCount);

  // Default direction for v1: PI/4 (NE)
  const direction = Math.PI / 4;

  return {
    sources: [{ direction, vertices, indices, vertexCount, indexCount }],
    sourceCount: 1,
    gridCols,
    gridRows,
    gridMinX,
    gridMinY,
    gridCellWidth,
    gridCellHeight,
  };
}

function parseV2(buffer: ArrayBuffer): WindMeshFileBundle {
  const view = new DataView(buffer);

  const sourceCount = view.getUint16(6, true);

  const dirTableStart = V2_HEADER_BYTES;
  const dirTableBytes = sourceCount * 4;
  const entryTableStart = dirTableStart + dirTableBytes;
  const entryTableBytes = sourceCount * 16;
  const gridMetaStart = entryTableStart + entryTableBytes;

  // Shared grid metadata
  const gridCols = view.getUint32(gridMetaStart, true);
  const gridRows = view.getUint32(gridMetaStart + 4, true);
  const gridMinX = view.getFloat32(gridMetaStart + 8, true);
  const gridMinY = view.getFloat32(gridMetaStart + 12, true);
  const gridCellWidth = view.getFloat32(gridMetaStart + 16, true);
  const gridCellHeight = view.getFloat32(gridMetaStart + 20, true);

  const sources: WindMeshSourceData[] = [];
  for (let i = 0; i < sourceCount; i++) {
    const direction = view.getFloat32(dirTableStart + i * 4, true);

    const entryOff = entryTableStart + i * 16;
    const vertexDataOffset = view.getUint32(entryOff, true);
    const vertexCount = view.getUint32(entryOff + 4, true);
    const indexDataOffset = view.getUint32(entryOff + 8, true);
    const indexCount = view.getUint32(entryOff + 12, true);

    const vertices = new Float32Array(
      buffer,
      vertexDataOffset,
      vertexCount * 5,
    );
    const indices = new Uint32Array(buffer, indexDataOffset, indexCount);

    sources.push({ direction, vertices, indices, vertexCount, indexCount });
  }

  return {
    sources,
    sourceCount,
    gridCols,
    gridRows,
    gridMinX,
    gridMinY,
    gridCellWidth,
    gridCellHeight,
  };
}

export function parseWindmeshBuffer(buffer: ArrayBuffer): WindMeshFileBundle {
  const view = new DataView(buffer);

  const magic = view.getUint32(0, true);
  if (magic !== MAGIC) {
    throw new Error(
      `Invalid windmesh magic: 0x${magic.toString(16)} (expected 0x${MAGIC.toString(16)})`,
    );
  }

  const version = view.getUint16(4, true);
  if (version === 1) {
    return parseV1(buffer);
  }
  if (version === 2) {
    return parseV2(buffer);
  }
  throw new Error(`Unsupported windmesh version: ${version}`);
}

File: pipeline/wavemesh-builder/src/windmesh_file.rs
use crate::level::TerrainCPUData;
use crate::windmesh::WindMeshData;

const MAGIC: u32 = 0x4d444e57; // "WNDM" little-endian
const VERSION: u16 = 2;
const HEADER_BYTES: usize = 32;

/// Build a version 2 multi-source windmesh buffer.
///
/// Format:
///   Header (32 bytes):
///     [0..3]   magic: "WNDM"
///     [4..5]   version: u16 = 2
///     [6..7]   sourceCount: u16
///     [8..15]  inputHash (2 x u32)
///     [16..31] reserved
///
///   Per-source direction table (4 bytes x sourceCount):
///     direction: f32 (radians)
///
///   Per-source entry table (16 bytes x sourceCount):
///     vertexDataOffset, vertexCount, indexDataOffset, indexCount (all u32)
///
///   Shared grid metadata (32 bytes):
///     gridCols, gridRows, gridMinX, gridMinY, gridCellWidth, gridCellHeight, reserved x2
///
///   Data sections (concatenated):
///     Vertex data per source (5 f32/vertex)
///     Index data (shared — same triangulation for all sources)
pub fn build_windmesh_buffer(meshes: &[WindMeshData], input_hash: [u32; 2]) -> Vec<u8> {
    let source_count = meshes.len();
    assert!(source_count > 0, "need at least one wind mesh");

    // All sources share the same grid dimensions and index data
    let ref_mesh = &meshes[0];
    let grid_cols = ref_mesh.grid_cols;
    let grid_rows = ref_mesh.grid_rows;
    let index_count = ref_mesh.index_count;

    let direction_table_bytes = source_count * 4;
    let entry_table_bytes = source_count * 16;
    let grid_metadata_bytes = 32;

    let per_source_vertex_bytes = ref_mesh.vertex_count * 5 * 4;
    let total_vertex_bytes = per_source_vertex_bytes * source_count;
    let index_data_bytes = index_count * 4; // shared

    let data_section_start =
        HEADER_BYTES + direction_table_bytes + entry_table_bytes + grid_metadata_bytes;
    let total_size = data_section_start + total_vertex_bytes + index_data_bytes;

    let mut buf = vec![0u8; total_size];

    // Header (32 bytes)
    buf[0..4].copy_from_slice(&MAGIC.to_le_bytes());
    buf[4..6].copy_from_slice(&VERSION.to_le_bytes());
    buf[6..8].copy_from_slice(&(source_count as u16).to_le_bytes());
    buf[8..12].copy_from_slice(&input_hash[0].to_le_bytes());
    buf[12..16].copy_from_slice(&input_hash[1].to_le_bytes());
    // [16..31] reserved = 0

    // Per-source direction table
    let dir_table_start = HEADER_BYTES;
    for (i, mesh) in meshes.iter().enumerate() {
        let off = dir_table_start + i * 4;
        buf[off..off + 4].copy_from_slice(&(mesh.direction as f32).to_le_bytes());
    }

    // Per-source entry table
    let entry_table_start = dir_table_start + direction_table_bytes;
    let index_data_offset = data_section_start + total_vertex_bytes;

    for (i, mesh) in meshes.iter().enumerate() {
        let off = entry_table_start + i * 16;
        let vertex_data_offset = data_section_start + i * per_source_vertex_bytes;
        buf[off..off + 4].copy_from_slice(&(vertex_data_offset as u32).to_le_bytes());
        buf[off + 4..off + 8].copy_from_slice(&(mesh.vertex_count as u32).to_le_bytes());
        buf[off + 8..off + 12].copy_from_slice(&(index_data_offset as u32).to_le_bytes());
        buf[off + 12..off + 16].copy_from_slice(&(mesh.index_count as u32).to_le_bytes());
    }

    // Shared grid metadata (32 bytes)
    let gm = entry_table_start + entry_table_bytes;
    buf[gm..gm + 4].copy_from_slice(&(grid_cols as u32).to_le_bytes());
    buf[gm + 4..gm + 8].copy_from_slice(&(grid_rows as u32).to_le_bytes());
    buf[gm + 8..gm + 12].copy_from_slice(&(ref_mesh.grid_min_x as f32).to_le_bytes());
    buf[gm + 12..gm + 16].copy_from_slice(&(ref_mesh.grid_min_y as f32).to_le_bytes());
    buf[gm + 16..gm + 20].copy_from_slice(&(ref_mesh.grid_cell_width as f32).to_le_bytes());
    buf[gm + 20..gm + 24].copy_from_slice(&(ref_mesh.grid_cell_height as f32).to_le_bytes());
    // [gm+24..gm+32] reserved = 0

    // Vertex data per source
    for (i, mesh) in meshes.iter().enumerate() {
        let start = data_section_start + i * per_source_vertex_bytes;
        for (j, &val) in mesh.vertices[..mesh.vertex_count * 5].iter().enumerate() {
            let off = start + j * 4;
            buf[off..off + 4].copy_from_slice(&val.to_le_bytes());
        }
    }

    // Index data (shared — use first mesh's indices)
    for (j, &val) in ref_mesh.indices[..index_count].iter().enumerate() {
        let off = index_data_offset + j * 4;
        buf[off..off + 4].copy_from_slice(&val.to_le_bytes());
    }

    buf
}

fn fnv1a_32(parts: &[&[u8]], offset_basis: u32) -> u32 {
    let mut h = offset_basis;
    for part in parts {
        for &b in *part {
            h ^= b as u32;
            h = h.wrapping_mul(0x01000193);
        }
    }
    h
}

fn f64_bytes(v: f64) -> [u8; 8] {
    v.to_le_bytes()
}

pub fn compute_wind_input_hash(
    terrain: &TerrainCPUData,
    wind_directions: &[f64],
) -> [u32; 2] {
    let vertex_bytes: Vec<u8> = terrain
        .vertex_data
        .iter()
        .flat_map(|f| f.to_le_bytes())
        .collect();
    let children_bytes: Vec<u8> = terrain
        .children_data
        .iter()
        .flat_map(|u| u.to_le_bytes())
        .collect();
    let contour_count_bytes = f64_bytes(terrain.contour_count as f64);
    let default_depth_bytes = f64_bytes(terrain.default_depth);
    let direction_bytes: Vec<u8> = wind_directions
        .iter()
        .flat_map(|d| d.to_le_bytes())
        .collect();

    let parts: Vec<Vec<u8>> = vec![
        vertex_bytes,
        terrain.contour_data.clone(),
        children_bytes,
        contour_count_bytes.to_vec(),
        default_depth_bytes.to_vec(),
        direction_bytes,
    ];

    let refs: Vec<&[u8]> = parts.iter().map(|p| p.as_slice()).collect();
    [fnv1a_32(&refs, 0x811c9dc5), fnv1a_32(&refs, 0x050c5d1f)]
}
